src/Microsoft.OData.Core/Json/JsonWriter.cs

```diff
@@ -621,6 +621,493 @@ public void EndTextWriterValueScope()
+        {
+            Debug.Assert(value.ValueKind == JsonValueKind.Object);
+
+            await this.StartObjectScopeAsync().ConfigureAwait(false);
+            foreach (JsonProperty property in value.EnumerateObject())
+            {
+                await this.WriteNameAsync(property.Name).ConfigureAwait(false);
+                await this.WriteValueAsync(property.Value).ConfigureAwait(false);
+            }
+
+            await this.EndObjectScopeAsync().ConfigureAwait(false);
+        }
+
+        private Task WriteJsonElementNumberAsync(JsonElement value)
+        {
+            Debug.Assert(value.ValueKind == JsonValueKind.Number);
+
+            if (value.TryGetByte(out byte byteValue))
+            {
+                return this.WriteValueAsync(byteValue);
+            }
+
+            if (value.TryGetDecimal(out decimal decimalValue))
+            {
+                return this.WriteValueAsync(decimalValue);
+            }
+
+            if (value.TryGetDouble(out double doubleValue))
+            {
+                return WriteValueAsync(doubleValue);
+            }
+
+            if (value.TryGetInt16(out short shortValue))
+            {
+                return this.WriteValueAsync(shortValue);
+            }
+
+            if (value.TryGetInt32(out int intValue))
+            {
+                return this.WriteValueAsync(intValue);
+            }
+
+            if (value.TryGetInt64(out long longValue))
+            {
+                return this.WriteValueAsync(longValue);
+            }
+
+            if (value.TryGetSByte(out sbyte sbyteValue))
+            {
+                return this.WriteValueAsync(sbyteValue);
+            }
+
+            if (value.TryGetSingle(out float floatValue))
+            {
+                return this.WriteValueAsync(floatValue);
+            }
+
+            if (value.TryGetUInt16(out ushort ushortValue))
+            {
+                return this.WriteValueAsync(ushortValue);
+            }
+
+            if (value.TryGetUInt32(out uint uintValue))
+            {
+                return this.WriteValueAsync(uintValue);
+            }
+
+            if (value.TryGetUInt64(out ulong ulongValue))
+            {
+                return this.WriteValueAsync((decimal)ulongValue);
+            }
+
+            Debug.Fail($"Exhausted all known JSON number types and did not find match.");
+            return Task.CompletedTask;
+        }
+#endif
+
+        /// <inheritdoc/>
+        public async Task WriteRawValueAsync(string rawValue)
+        {
+            await this.WriteValueSeparatorAsync().ConfigureAwait(false);
+            await this.writer.WriteAsync(rawValue).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public Task FlushAsync()
+        {
+            return this.writer.FlushAsync();
+        }
+
+#if NETCOREAPP3_1_OR_GREATER
+        public async ValueTask DisposeAsync()
+        {
+            if (this.ArrayPool != null && this.wrappedBuffer.Value != null)

```

Comment:
Unrelated, but I think we should get rid of the `ArrayPool` property as well as our custom `ICharArrayPool` interface and use the built-in array pool instead. I'll create an issue to discuss this potentially for ODL 9.

Reply:
@habbes An issue to track that would be good